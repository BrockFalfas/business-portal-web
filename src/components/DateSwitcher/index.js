import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Button, Icon } from 'antd';
import classNames from 'classnames';
import moment from 'moment';

export const addToDate = (date, interval, value) => {
  const intervalName = `${interval}s`;
  return moment(date).add(value, intervalName);
};

export default class DateSwitcher extends Component {
  static propTypes = {
    className: PropTypes.string,
    date: PropTypes.object,
    interval: PropTypes.oneOf(['day', 'month', 'year']),
  };

  static defaultProps = {
    date: moment(),
    interval: 'month',
  };

  static dateFormats = {
    day: 'MM/DD/YYYY',
    month: 'MMM YYYY',
    year: 'YYYY',
  };

  constructor(props) {
    super(props);

    this.addToDate = addToDate;
  }

  render() {
    const { className } = this.props;

    return (
      <div className={classNames('DateSwitcher', className)}>
        <Button onClick={this.handleDateDecrease}>
          <Icon type="left" />
        </Button>
        <div className="DateSwitcher-value">{this.formatDate()}</div>
        <Button onClick={this.handleDateIncrease}>
          <Icon type="right" />
        </Button>
      </div>
    );
  }

  handleDateIncrease = () => {
    const { date, interval } = this.props;
    const newDate = this.addToDate(date, interval, 1);
    this.handleChange(newDate);
  };

  handleDateDecrease = () => {
    const { date, interval } = this.props;
    const newDate = this.addToDate(date, interval, -1);
    this.handleChange(newDate);
  };

  handleChange = newDate => {
    this.setState({ date: newDate });
    this.props.onChange(newDate);
  };

  formatDate = () => {
    const { date, interval } = this.props;
    return date.format(DateSwitcher.dateFormats[interval]);
  };
}
