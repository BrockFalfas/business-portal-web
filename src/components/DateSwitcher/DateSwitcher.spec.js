import React from 'react';
import { shallow } from 'enzyme';
import moment from 'moment';

import DateSwitcher, { addToDate } from './index';

const initDateSwitcher = overrides => {
  const mockProps = {
    onChange: jest.fn(),
  };
  const wrapper = shallow(<DateSwitcher {...mockProps} {...overrides} />);

  return { mockProps, wrapper };
};

describe('component: <DateSwitcher />', () => {
  describe('rendering', () => {
    it('should render without crashing', () => {
      const { wrapper } = initDateSwitcher();
      expect(wrapper).toMatchSnapshot();
    });

    it('should properly format date based on interval', () => {
      const mockDate = moment('2018-08-01T00:00:00Z');
      const { wrapper } = initDateSwitcher({ interval: 'day', date: mockDate });
      expect(wrapper.find('.DateSwitcher-value').text()).toBe(
        mockDate.format(DateSwitcher.dateFormats['day'])
      );

      wrapper.setProps({ interval: 'month' });
      expect(wrapper.find('.DateSwitcher-value').text()).toBe(
        mockDate.format(DateSwitcher.dateFormats['month'])
      );

      wrapper.setProps({ interval: 'year' });
      expect(wrapper.find('.DateSwitcher-value').text()).toBe(
        mockDate.format(DateSwitcher.dateFormats['year'])
      );
    });
  });

  describe('methods', () => {
    describe('formatDate', () => {
      it('should format the date using predefined format for provided interval', () => {
        const mockDate = moment('2018-08-01T00:00:00Z');
        mockDate.format = jest.fn();
        const { wrapper } = initDateSwitcher({ date: mockDate });

        Object.entries(DateSwitcher.dateFormats).forEach(([key, value]) => {
          wrapper.setProps({ interval: key });
          wrapper.instance().formatDate();
          expect(mockDate.format).toHaveBeenLastCalledWith(value);
        });
      });
    });

    describe('handleChange', () => {
      it('should update the date in the state and call onChange event', () => {
        const { mockProps, wrapper } = initDateSwitcher();
        wrapper.instance().handleChange('fake date');
        expect(wrapper.state('date')).toBe('fake date');
        expect(mockProps.onChange).toHaveBeenCalledWith('fake date');
      });
    });

    describe('handleDateIncrease', () => {
      it('should increase the date by interval and call handleChange', () => {
        const mockDate = moment();
        const { wrapper } = initDateSwitcher({ date: mockDate });
        const instance = wrapper.instance();
        instance.addToDate = jest.fn(() => 'fake date');
        instance.handleChange = jest.fn();

        Object.keys(DateSwitcher.dateFormats).forEach(key => {
          wrapper.setProps({ interval: key });
          instance.handleDateIncrease();
          expect(instance.addToDate).toHaveBeenLastCalledWith(mockDate, key, 1);
          expect(instance.handleChange).toHaveBeenCalledWith('fake date');
        });
      });
    });

    describe('handleDateDecrease', () => {
      it('should increase the date by interval and call handleChange', () => {
        const mockDate = moment();
        const { wrapper } = initDateSwitcher({ date: mockDate });
        const instance = wrapper.instance();
        instance.addToDate = jest.fn(() => 'fake date');
        instance.handleChange = jest.fn();

        Object.keys(DateSwitcher.dateFormats).forEach(key => {
          wrapper.setProps({ interval: key });
          instance.handleDateDecrease();
          expect(instance.addToDate).toHaveBeenLastCalledWith(mockDate, key, -1);
          expect(instance.handleChange).toHaveBeenCalledWith('fake date');
        });
      });
    });
  });

  describe('addToDate', () => {
    let mockDate;
    const formatDate = date => date.utc().format();

    beforeEach(() => {
      mockDate = moment('2018-08-01T00:00:00Z');
    });

    it('should properly add and subtract days', () => {
      expect(formatDate(addToDate(mockDate, 'day', 2))).toBe('2018-08-03T00:00:00Z');
      expect(formatDate(addToDate(mockDate, 'day', -2))).toBe('2018-07-30T00:00:00Z');
    });

    it('should properly add and subtract months', () => {
      expect(formatDate(addToDate(mockDate, 'month', 2))).toBe('2018-10-01T00:00:00Z');
      expect(formatDate(addToDate(mockDate, 'month', -2))).toBe('2018-06-01T00:00:00Z');
    });

    it('should properly add and subtract years', () => {
      expect(formatDate(addToDate(mockDate, 'year', 2))).toBe('2020-08-01T00:00:00Z');
      expect(formatDate(addToDate(mockDate, 'year', -2))).toBe('2016-08-01T00:00:00Z');
    });
  });
});
