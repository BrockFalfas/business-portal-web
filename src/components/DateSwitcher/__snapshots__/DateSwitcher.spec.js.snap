// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: <DateSwitcher /> rendering should render without crashing 1`] = `
<div
  className="DateSwitcher"
>
  <Button
    onClick={[Function]}
  >
    <Icon
      type="left"
    />
  </Button>
  <div
    className="DateSwitcher-value"
  >
    Aug 2018
  </div>
  <Button
    onClick={[Function]}
  >
    <Icon
      type="right"
    />
  </Button>
</div>
`;
